### PR TITLE
[DRAFT][RHCLOUD-44786] Update role implementation

### PR DESCRIPTION
**Pull request by Ellen-Yi-Dong**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-44786

## Description of Intent of Change(s)
The what, why and how.
- add the update role implementation (write) based on the criteria listed within the JIRA ticket 
- currently waiting for the outbox replication 

## Local Testing
How can the feature be exercised?
How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

## Summary by Sourcery

Implement role update functionality for v2 roles, including service-layer support, API endpoint handling, and field-selectable responses that align with the existing create behavior.

New Features:
- Add v2 role update operation in the service layer to modify name, description, and permissions for an existing role.
- Expose a PUT /api/v2/roles/{uuid}/ endpoint that updates roles via the serializer and returns a field-selectable response payload consistent with the create endpoint.

Enhancements:
- Refactor role creation validation into a shared helper to reuse description and permission validation for both create and update flows.
- Introduce a dedicated RoleNotFoundError domain exception and map domain errors to appropriate API-level responses in the serializer.
- Add support for a fields query parameter on create/update responses, with sensible defaults and validation of requested fields.

Tests:
- Add unit tests for role update behavior in the service layer, including validation, duplicate name handling, and non-existent role cases.
- Extend API view tests to cover role update success and error scenarios, as well as fields query parameter handling and default field behavior.
- Add serializer tests to verify update delegation to the service and correct error translation for not-found and duplicate-name situations.

### CODE REVIEW

**Comment by lpichler** [comment]
@Ellen-Yi-Dong can you rebase ?

File: rbac/management/role/v2_view.py [hunk 1]
```diff
@@ -69,10 +74,40 @@ def get_serializer_context(self):
 
     def get_serializer_class(self):
         """Return appropriate serializer based on action."""
-        if self.action == "create":
+        if self.action in ("create", "update"):
             return RoleV2RequestSerializer
         return RoleV2ResponseSerializer
 
+    def _parse_fields_param(self, fields_str: str | None) -> set[str]:
```
Comment (lennysgarage) [inline rbac/management/role/v2_view.py]: I'm wondering if this whole function is required or not, for list I was using the RoleFieldSelection class directly via a `validate_fields` function that exists on the response serializer. This way the view doesn't need to worry at all about the fields params. The response serializer via `is_valid()` will call out and validate the params and raise the exception if invalid. [food for thought](https://github.com/RedHatInsights/insights-rbac/blob/master/rbac/management/role/v2_view.py#L78-L80) 
```
    def validate_fields(self, value):
        """Parse, validate, and resolve fields parameter into a set of field names."""
        return _validate_fields_parameter(value, RoleV2Service.DEFAULT_LIST_FIELDS)
```

That way the serializer can handle the error handling

File: rbac/management/role/v2_exceptions.py [hunk 2]
```diff
@@ -64,3 +55,12 @@ class InvalidRolePermissionsError(RoleV2Error):
     def __init__(self, message: str):
         """Initialize with the validation error message."""
         super().__init__(message)
+
+
+class RoleNotFoundError(RoleV2Error):
```
Comment (lennysgarage) [inline rbac/management/role/v2_exceptions.py]: nit.

but do we need to move this class around? seems like the changes are the same and prob from a merge

File: rbac/management/role/v2_service.py [hunk 3]
```diff
@@ -144,6 +156,80 @@ def create(
...
+            # Update role fields
+            role.name = name
+            role.description = description
+            role.save()
+            role.permissions.set(permissions)
+
+            # Capture new state after update for outbox replication
+            # Django's .set() clears the cached queryset, so .as_tuples() will
+            # query fresh permissions from the database
+            new_tuples = role.as_tuples()
```
Comment (astrozzc) [inline rbac/management/role/v2_service.py]: I wonder if we can update the as_tuples method to check if role.permissions are set e.g., role.permissions.set(permissions), if it is use role.permissions directly instead of another role.permissions.all() query

File: rbac/management/role/v2_service.py [hunk 4]
```diff
@@ -144,6 +156,80 @@ def create(
...
+                ReplicationEvent(
+                    event_type=ReplicationEventType.UPDATE_CUSTOM_ROLE,
+                    info={"role_uuid": str(role.uuid), "org_id": str(tenant.org_id)},
+                    partition_key=PartitionKey.byEnvironment(),
+                    remove=old_tuples,
+                    add=new_tuples,
+                )
+            )
+
+            logger.info(
```
Comment (snehagunta) [inline rbac/management/role/v2_service.py]: Two suggestions here:

**1. Use `CustomRoleV2.replication_tuples` which encapsulates the full list of tuplesToCreate and tuplesToRemove and can be reused by all the APIs

**2. Add an `update()` method to `CustomRoleV2`**

Add to `CustomRoleV2` in `v2_model.py`:
```python
def update(self, name: str, description: str):
    """Update the role's mutable attributes."""
    self.name = name
    self.description = description
```

Then the service becomes:
```suggestion
            old_permissions = list(role.permissions.all())

            role.update(name, description)
            role.save()
            role.permissions.set(permissions)

            tuples_to_add, tuples_to_remove = CustomRoleV2.replication_tuples(
                role, old_permissions=old_permissions, new_permissions=permissions
            )

            self._replicator.replicate(
                ReplicationEvent(
                    event_type=ReplicationEventType.UPDATE_CUSTOM_ROLE,
                    info={"role_uuid": str(role.uuid), "org_id": str(tenant.org_id)},
                    partition_key=PartitionKey.byEnvironment(),
                    add=tuples_to_add,
                    remove=tuples_to_remove,
                )
            )
```

**Review by snehagunta — COMMENTED** [review]
Suggestions for using `CustomRoleV2.replication_tuples` (from the `consolidate-role-replication` branch) and adding an `update()` method to `CustomRoleV2` for hexagonal architecture consistency.

**Review by snehagunta — APPROVED** [review]
